docs/fundamentals/code-analysis/quality-rules/ca1841.md

```diff
@@ -0,0 +1,100 @@
+---
+title: "CA1841: Prefer Dictionary contains methods (code analysis)"
+description: "Learn about code analysis rule CA1841: Prefer Dictionary contains methods"
+ms.date: 04/21/2021
+ms.topic: reference
+f1_keywords:
+  - "PreferDictionaryContainsMethods"
+  - "CA1841"
+helpviewer_keywords:
+  - "PreferDictionaryContainsMethods"
+  - "CA1841"
+author: NewellClark
+ms.author: NewellClark
+dev_langs:
+  - CSharp
+  - VB
+---
+# CA1841: Prefer Dictionary contains methods
+
+| | Value |
+|-|-|
+| **Rule ID** |CA1841
+| **Category** |[Performance](performance-warnings.md)|
+| **Fix is breaking or non-breaking** |Non-breaking|
+
+## Cause
+
+This rule locates calls to a `Contains` method on the `Keys` or `Values` collection of an <xref:System.Collections.Generic.IDictionary%602> that could be replaced with a call to an accessible `ContainsKey` or `ContainsValue` method on the dictionary itself.
+
+## Rule description
+
+Calling `Contains` on the `Keys` or `Values` collection may often be more expensive than simply calling `ContainsKey` or `ContainsValue` on the dictionary itself:
+
+- Many dictionary implementations lazily instantiate the key and value collections, which means that accessing the `Keys` or `Values` collection may result in extra allocations.
+- You may end up calling an extension method on <xref:System.Collections.Generic.IEnumerable%601> if the keys or values collection uses explicit interface implementation to hide methods on <xref:System.Collections.Generic.ICollection%601>. This can lead to reduced performance, especially when accessing the key collection, as most dictionary implementations are able to provide a fast O(1) containment check for keys, while the `Contains` extension method on <xref:System.Collection.Generic.IEnumerable%601> usually does a slow O(n) containment check.
```

Comment:
```suggestion
- You may end up calling an extension method on <xref:System.Collections.Generic.IEnumerable%601> if the keys or values collection uses explicit interface implementation to hide methods on <xref:System.Collections.Generic.ICollection%601>. This can lead to reduced performance, especially when accessing the key collection, as most dictionary implementations are able to provide a fast O(1) containment check for keys, while the `Contains` extension method on <xref:System.Collections.Generic.IEnumerable%601> usually does a slow O(n) containment check.
```